course3/homework-unique_ptr/test.cc

```diff
@@ -0,0 +1,99 @@
+#include "unique_ptr.h"
+#include "unique_ptr.cc"
+
+#include <cassert>
+#include <iostream>
+#include <memory>
+
+
+template<template<class...> class UniquePtrType>
+void unique_ptr_test()
+{
+	// default contructor
+	UniquePtrType<int> ptr;
+	assert(!ptr);
+	assert(ptr.get() == nullptr);
+	
+	// initialization constructor
+	UniquePtrType<int> ptr2(nullptr);
+	assert(!ptr2);
+	
+	int* i = new int(42);
+	UniquePtrType<int> ptr3(i);
+	assert(ptr3);
+	assert(*ptr3.get() == 42);
+	
+	// move constructor
+	UniquePtrType<int> ptr4 = std::move(ptr3);
+	assert(*ptr4.get() == 42);
+	assert(ptr4.get() == i);
+	assert(!ptr3);
+	
+	// move assignment operator
+	ptr3 = std::move(ptr4);
+	assert(*ptr3.get() == 42);
+	assert(ptr3.get() == i);
+	assert(!ptr4);
+	
+	// nullptr assignment operator
+	ptr3 = nullptr;
+	assert(ptr3.get() == nullptr);
+	
+	// reset
+	ptr.reset(new int(42));
+	assert(*ptr.get() == 42);
+	
+	ptr.reset();
+	assert(!ptr);
+	
+	ptr.reset(new int(42));
+	ptr.reset(nullptr);
+	assert(!ptr);
+	
+	// release
+	ptr.reset(new int(42));
+	i = ptr.release();
+	assert(*i == 42);
+	assert(!ptr);
+	
+	ptr.reset(i);
+	assert(*ptr.get() == 42);
+	
+	ptr.reset();
+	i = ptr.release();
+	assert(i == nullptr);
+	assert(!ptr);
+	
+	// swap
+	ptr2.reset(new int(42));
+	ptr.swap(ptr2);
+	assert(!ptr2);
+	assert(*ptr.get() == 42);
+	
+	// operator== and operator!=
+	assert(ptr2 == ptr3);
+	assert(ptr != ptr2);
+	
+	// operator*
+	struct S { int i = 42; };
+	UniquePtrType<S> ptr5(new S);
+	S s = *ptr5;
+	assert(s.i == 42);
+	
+	// operator->
+	assert(ptr5->i == 42);
+	
```

Comment:
You don't really need to check whether the memory was freed, you need t check if 'delete' has been called on the object, because calling 'delete' is the only responsibility of the programmer who used 'new'.

How I usually do it, I have a small class that has a counter as a static member variable, and I increase the counter in constructors (default and copy), and decrease it in destructor. This way you can check how many instances currently exists.